src/os/bluestore/BlueStore.cc

```diff
@@ -10878,41 +10895,30 @@ int BlueStore::_collection_list(
 	dout(30) << __func__ << " pend " << pend << dendl;
 	continue;
       }
-      if (it->valid() && it->is_lt(coll_range_end)) {
-        *pnext = it->oid();
-        set_next = true;
+      if (it->valid()) {
+        int cmp_result = it->cmp(coll_range_end);
+        if (cmp_result < 0) {
+          *pnext = it->oid();
+        } else if (cmp_result > 0) {
+          *pnext = ghobject_t::get_max();
+	} else {
+          ls->push_back(it->oid());
+          *pnext = ghobject_t::get_max();
```

Comment:
not quite right, if the iterator is not hitting the 0xffffffff upper bound, in that case, we should not set the returning "next" pointer to max. 

Probably, should just take the 0xfffffff as an exception and compare the iterator with it and `pend`.

Reply:
Just thinking, what if instead of complicating this by adding this special handling we make sure `coll_range_temp_end` and `coll_range_end` are "the biggest" oids. I.e. modify `get_coll_range` to set not only the pool and the bitwise key, but also the namespace to some "MAX" value?

It might be a bit hackish to choose this "MAX" value for a string, but probably filling it with 0xff bytes (like `std::string(0xff, 8)`) would be good enough?

Reply:
Eh, I tried it and it does not seem like so simple. I have started to observe errors like
```
2021-09-24T08:08:29.580+0100 7f9630121640  0 log_channel(cluster) log [WRN] : 1.25 found stray pgmeta-like #1:a8000000::::head# during PG removal
```
Which I think are related.

Reply:
Thinking a little more about it. I suppose I had to modify `coll_range_temp_end` to set "MAX" namespace only for the case when `end_hash == 0xffffffffull`.  For the `end_hash` we have:
```
    uint64_t end_hash = reverse_hash  + (1ull << (32 - bits));
    if (end_hash > 0xffffffffull)
      end_hash = 0xffffffffull;
```
I.e. normally `end_hash` is out of the range, and `end` oid built with this hash will be always the biggests. But when we overflow (`end_hash` becomes `0xffffffffull + 1`) we set it to 0xffffffffull and then have this weird things.

At least this is my interpretation. And I tried this patch and it seems to work for me (though not tested well):
```
--- a/src/os/bluestore/BlueStore.cc
+++ b/src/os/bluestore/BlueStore.cc
@@ -311,6 +311,11 @@ static void get_coll_range(const coll_t& cid, int bits,
 
     end->hobj.set_bitwise_key_u32(end_hash);
     temp_end->hobj.set_bitwise_key_u32(end_hash);
+
+    if (end_hash == 0xffffffffull) {
+      end->hobj.nspace = std::string(0xff, 8);
+      temp_end->hobj.nspace = std::string(0xff, 8);
+    }
   } else {
     start->shard_id = shard_id_t::NO_SHARD;
     start->hobj.pool = -1ull;
@@ -322,6 +327,9 @@ static void get_coll_range(const coll_t& cid, int bits,
     // no separate temp section
     *temp_start = *end;
     *temp_end = *end;
+
+    end->hobj.nspace = std::string(0xff, 8);
+    temp_end->hobj.nspace = std::string(0xff, 8);
   }
 
   start->generation = 0;
```
And I am not telling this is the right solution.

Reply:
i think this will work, one little nit, why std::string(0xff, 8), should  std::string(0xff, 4) instead？

Reply:
`std::string(0xff, 8)` was just the first thing that came to my mind. I suppose even just `std::string(0xff, 1)`  would be enough.